src/WRings.sol

```diff
@@ -0,0 +1,124 @@
+        exchangeRate = exchangeRate_;
+    }
+
+    /// @notice The amount of shares to mint to a user for the asset token deposited under ideal conditions
+    /// @param assets Number of Rings tokens to deposit to the contract 
+    /// @return shares Number of vault shares minted to the user based on the exchange rate
+    function convertToShares(uint256 assets) external view returns(uint256) {
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        
+        return shares;
+    }
+
+    /// @notice Used to calculate the amount of an asset token to transfer to user for vault tokens burned under ideal conditions
+    /// @param shares Number of vault tokens to burn
+    /// @return assets Number of asset tokens to transfer to the user
+    function convertToAssets(uint256 shares) external view returns(uint256) {
+        uint256 assets = WDiv.wdiv(shares * 10**9, exchangeRate) / 10**9;
+        return assets;
+    }
+
+    /// @notice User specifies amount of asset token to deposit in return for vault shares
+    /// @param assets Number of asset tokens to deposit
+    /// @param receiver Address that vault tokens will be minted to
+    /// @return shares Number of vault shares to return to the user
+    function deposit(uint256 assets, address receiver) external  returns(uint256) {
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        mint(receiver, shares);
+        iRings.transferFrom(msg.sender, address(this), assets);
+        emit Deposit(msg.sender, receiver, assets, shares);
+        return shares;
+    }
+
+    /// @notice User specifies the desired amount of vault shares in exchange for asset tokens
+    /// @param shares Number of vault tokens desired by the user
+    /// @param receiver Address that vault Tokens will be minted to
+    /// @return assets Number of asset tokens required to mint desired amount of shares
+    function mint(uint256 shares, address receiver) external  returns(uint256) {
+        uint256 assets = WDiv.wdiv(shares * 10**9, exchangeRate) / 10**9;
+        mint(receiver, shares);
+        iRings.transferFrom(msg.sender, address(this), assets);
+        emit Deposit(msg.sender, receiver, assets, shares);
+        return assets;
+    }
+
+    /// @notice User specifies amount of asset tokens to withdraw from the vault
+    /// @dev
+    /// @param assets Number of asset token that user desires to withdraw from the vault
+    /// @param receiver Address that will receive the withdrawn tokens
+    /// @param owner Address that currently owns the vault tokens
+    /// @return shares Number of shares required to be exchange for desired amount of asset tokens
+    function withdraw(uint256 assets, address receiver, address owner) external  returns(uint256){
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
+        require(_balanceOf[owner] >= shares, "Owner holds less than required amount of vault shares for requested assets");
```

Comment:
Error messages should be shorter than 32 characters, by convention and to save deployment gas.

Reply:
Thank you, never heard that anywhere before